File: backend-spring-boot/src/main/java/no/vicx/backend/user/validation/AtLeastOneNotNullValidator.java
package no.vicx.backend.user.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import no.vicx.backend.user.vm.UserPatchVm;

public class AtLeastOneNotNullValidator implements ConstraintValidator<AtLeastOneNotNull, UserPatchVm> {

    private String defaultMessage;
    private String propertyNodeName;

    @Override
    public void initialize(AtLeastOneNotNull constraintAnnotation) {
        defaultMessage = constraintAnnotation.message();
        propertyNodeName = constraintAnnotation.propertyNodeName();
    }

    @Override
    public boolean isValid(UserPatchVm value, ConstraintValidatorContext context) {
        if (value.isEmpty()) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate(defaultMessage)
                    .addPropertyNode(propertyNodeName)
                    .addConstraintViolation();

            return false;
        }

        return true;
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/user/validation/AtLeastOneNotNullValidator.kt
package no.vicx.backend.user.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import no.vicx.backend.user.vm.UserPatchVm

class AtLeastOneNotNullValidator : ConstraintValidator<AtLeastOneNotNull, UserPatchVm> {
    private var defaultMessage: String = ""
    private var propertyNodeName: String = ""

    override fun initialize(constraintAnnotation: AtLeastOneNotNull) {
        defaultMessage = constraintAnnotation.message
        propertyNodeName = constraintAnnotation.propertyNodeName
    }

    override fun isValid(
        value: UserPatchVm,
        context: ConstraintValidatorContext,
    ): Boolean {
        if (value.isEmpty) {
            context.disableDefaultConstraintViolation()
            context
                .buildConstraintViolationWithTemplate(defaultMessage)
                .addPropertyNode(propertyNodeName)
                .addConstraintViolation()

            return false
        }

        return true
    }
}
